### ISSUE TITLE
OptionalUnit: False positive in lambda

### ISSUE THREAD

**Issue by realdadfish** [arturbosch/detekt#1835]
This is detekt-1.0.0-RC14

Consider the following example where Mockito-Kotlin is used to stub an answer to a (unit/void) method call:

```kotlin
interface Foo {
  fun bar()
}

fun bla() {
  mock<Foo> {
    on { bar() } doAnswer {
      ignoredReturnValue()
      Unit
    }
  }
}

private fun ignoredReturnValue(): Int = 3
```

Mockito-Kotlin's `doAnswer` infix function requires a lambda that returns a `Unit?`. Since the last regular method call of the lambda however returns an `Int`, one has to write `Unit` to satisfy this dependency.
This `Unit` is then however marked by Detekt's `OptionalUnit` rule as unneccessary, though - if missing - prevents Kotlin from compiling the above code.

**Comment by schalkms** [comment]
I think this one would be very hard to detect, even if we use type and symbol solving. For me this is a classical `@Suppress` example.

**Comment by realdadfish** [comment]
Hrm... I don't know the PSI API good enough, but would it be somehow possible if you're on instruction X (which would be in this case `Unit`) that you could "look-behind" instruction X-1, basically the previous sibling?
Because then - in theory - you could check if this instruction returns a non-`Unit` type and if so, leave the `Unit` where it is because it would be needed in this case, no?

**Comment by schalkms** [comment]
> Because then - in theory - you could check if this instruction returns a non-Unit type and if so, leave the Unit where it is because it would be needed in this case, no?

This assumption is just valid for this case but not for other examples like functions or a simple `foreach { }`. Frankly, if you used this once I would go with the `@Suppress` annotation. If you used this multiple times, you could think of `exclude` patterns.

PS: Please update to detekt v1.0.0. Since RC14 a lot has changed.

**Comment by schalkms** [comment]
@arturbosch Type and symbol solving doesn’t help us here, because how would you just check the flagged code snippet and nothing more. I’d close this and consider it as a false-positive where we cannot do anything.

**Comment by arturbosch** [comment]
You're right, I was like "look at doAnswer and the lambda return type and match Unit".
